// @rnx-kit/tools-language/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pickValues = exports.pickValue = exports.hasProperty = exports.extendObjectArray = exports.extendObject = exports.isApproximatelyEqual = exports.tryInvoke = exports.addRange = exports.toIndex = void 0;
var array_1 = require("./array");
Object.defineProperty(exports, "toIndex", { enumerable: true, get: function () { return array_1.toIndex; } });
Object.defineProperty(exports, "addRange", { enumerable: true, get: function () { return array_1.addRange; } });
var function_1 = require("./function");
Object.defineProperty(exports, "tryInvoke", { enumerable: true, get: function () { return function_1.tryInvoke; } });
var math_1 = require("./math");
Object.defineProperty(exports, "isApproximatelyEqual", { enumerable: true, get: function () { return math_1.isApproximatelyEqual; } });
var properties_1 = require("./properties");
Object.defineProperty(exports, "extendObject", { enumerable: true, get: function () { return properties_1.extendObject; } });
Object.defineProperty(exports, "extendObjectArray", { enumerable: true, get: function () { return properties_1.extendObjectArray; } });
Object.defineProperty(exports, "hasProperty", { enumerable: true, get: function () { return properties_1.hasProperty; } });
Object.defineProperty(exports, "pickValue", { enumerable: true, get: function () { return properties_1.pickValue; } });
Object.defineProperty(exports, "pickValues", { enumerable: true, get: function () { return properties_1.pickValues; } });
